### ISSUE TITLE
Batching of URLs is inefficient

### ISSUE THREAD

**Issue by okdana** [apt-mirror/apt-mirror#80]
If i'm not mistaken, `apt-mirror` works by splitting the files to be downloaded up into batches of `files / nthreads` files each. So if you've got 1000 files to download, and you've configured 2 threads, thread `0` gets 500 files and thread `1` gets 500 files.

The problem with this arrangement is that each set of files may be vastly different in size. For example, thread `0` might get 500 small command-line utilities and thread `1` might get 500 giant kernel images. So what ends up happening here is that all of your threads except one complete, and then you're left sitting there waiting for the last one to slowly pull down its massive pay-load all by itself. It gets especially bad if you have a per-thread download limit set — you could be waiting *days* for that one lonely instance of `wget` to finish.

The package lists contain the size of each of the packages, right? Would it be possible to change the apportioning of files so that each `wget` instance gets a roughly similar *amount of data* (rather than *number of files*) to fetch?

**Comment by bdrung** [comment]
I agree that the batching of URLs is inefficient. Instead of splitting the files to download for each thread, a shared pool of not-yet-downloaded files should be used instead. Only when this pool is exhausted, the threads should exit.

I plan to rewrite apt-mirror in Python (no ETA) and would implement this change there.

**Comment by okdana** [comment]
Sounds good, thanks!
